src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+    association: u32,
+    sequence: u32,
+}
+
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+struct VirtIOSndPcmHdr {
+    /// specifies request type (VIRTIO_SND_R_PCM_*)
+    hdr: VirtIOSndHdr,
+    /// specifies a PCM stream identifier from 0 to streams - 1
+    stream_id: u32,
+}
+
+enum PcmStreamFeatures {
+    VirtioSndPcmFShmemHost = 0,
+    VirtioSndPcmFShmemGuest,
+    VirtioSndPcmFMsgPolling,
+    VirtioSndPcmFEvtShmemPeriods,
+    VirtioSndPcmFEvtXruns,
+}
+
+impl From<PcmStreamFeatures> for u32 {
+    fn from(value: PcmStreamFeatures) -> Self {
+        value as _
+    }
+}
+
+enum PcmSampleFormat {
+    /* analog formats (width / physical width) */
+    VirtioSndPcmFmtImaAdpcm = 0, /* 4 / 4 bits */
+    VirtioSndPcmFmtMuLaw,        /* 8 / 8 bits */
+    VirtioSndPcmFmtALaw,         /* 8 / 8 bits */
+    VirtioSndPcmFmtS8,           /* 8 / 8 bits */
+    VirtioSndPcmFmtU8,           /* 8 / 8 bits */
+    VirtioSndPcmFmtS16,          /* 16 / 16 bits */
+    VirtioSndPcmFmtU16,          /* 16 / 16 bits */
+    VirtioSndPcmFmtS18_3,        /* 18 / 24 bits */
+    VirtioSndPcmFmtU18_3,        /* 18 / 24 bits */
+    VirtioSndPcmFmtS20_3,        /* 20 / 24 bits */
+    VirtioSndPcmFmtU20_3,        /* 20 / 24 bits */
+    VirtioSndPcmFmtS24_3,        /* 24 / 24 bits */
+    VirtioSndPcmFmtU24_3,        /* 24 / 24 bits */
+    VirtioSndPcmFmtS20,          /* 20 / 32 bits */
+    VirtioSndPcmFmtU20,          /* 20 / 32 bits */
+    VirtioSndPcmFmtS24,          /* 24 / 32 bits */
+    VirtioSndPcmFmtU24,          /* 24 / 32 bits */
+    VirtioSndPcmFmtS32,          /* 32 / 32 bits */
+    VirtioSndPcmFmtU32,          /* 32 / 32 bits */
+    VirtioSndPcmFmtFloat,        /* 32 / 32 bits */
+    VirtioSndPcmFmtFloat64,      /* 64 / 64 bits */
+    /* digital formats (width / physical width) */
+    VirtioSndPcmFmtDsdU8,          /* 8 / 8 bits */
+    VirtioSndPcmFmtDsdU16,         /* 16 / 16 bits */
+    VirtioSndPcmFmtDsdU32,         /* 32 / 32 bits */
+    VirtioSndPcmFmtIec958Subframe, /* 32 / 32 bits */
+}
+
+impl From<PcmSampleFormat> for u64 {
+    fn from(value: PcmSampleFormat) -> Self {
+        value as _
+    }
+}
+
+enum PcmFrameRate {
+    VirtioSndPcmRate5512 = 0,
+    VirtioSndPcmRate8000,
+    VirtioSndPcmRate11025,
+    VirtioSndPcmRate16000,
+    VirtioSndPcmRate22050,
+    VirtioSndPcmRate32000,
+    VirtioSndPcmRate44100,
+    VirtioSndPcmRate48000,
+    VirtioSndPcmRate64000,
+    VirtioSndPcmRate88200,
+    VirtioSndPcmRate96000,
+    VirtioSndPcmRate176400,
+    VirtioSndPcmRate192000,
+    VirtioSndPcmRate384000,
+}
+
+impl Into<u64> for PcmFrameRate {
+    fn into(self) -> u64 {
+        1 << self as usize
+    }
+}
+
+/// PCM information.
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+pub struct VirtIOSndPcmInfo {
+    hdr: VirtIOSndInfo,
+    features: u32, /* 1 << VIRTIO_SND_PCM_F_XXX */
+    formats: u64,  /* 1 << VIRTIO_SND_PCM_FMT_XXX */
+    rate: u64,     /* 1 << VIRTIO_SND_PCM_RATE_XXX */
```

Comment:
Could this be a `PcmRate` rather than a `u64`, to avoid manually converting back and forth? Likewise the other bit flags fields.

Reply:
I believe we should not convert u32, u64, and so on into bitflags for these structure fields. The naming and types of these fields are copied directly from the C implementation in Virtio v1.2. Therefore, I think we should not modify these standard-defined structures.